zrml/hybrid-router/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+        /// * `who` - The account ID of the user performing the transaction.
+        /// * `market_id` - The ID of the market.
+        /// * `asset` - The asset to be traded.
+        /// * `amount_in` - The amount to be traded.
+        /// * `price_limit` - The maximum or minimum price at which the trade can be executed.
+        fn maybe_fill_from_amm(
+            tx_type: TxType,
+            who: &AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            asset: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            price_limit: BalanceOf<T>,
+        ) -> Result<(BalanceOf<T>, Option<AmmTradeOf<T>>), DispatchError> {
+            if !T::Amm::pool_exists(market_id) {
+                return Ok((amount_in, None));
+            }
+
+            let spot_price = T::Amm::get_spot_price(market_id, asset)?;
+
+            let amm_amount_in = match tx_type {
+                TxType::Buy => {
+                    if spot_price >= price_limit {
+                        return Ok((amount_in, None));
+                    }
+                    T::Amm::calculate_buy_amount_until(market_id, asset, price_limit)?
+                }
+                TxType::Sell => {
+                    if spot_price <= price_limit {
+                        return Ok((amount_in, None));
+                    }
+                    T::Amm::calculate_sell_amount_until(market_id, asset, price_limit)?
+                }
+            };
+
+            let amm_amount_in = amm_amount_in.min(amount_in);
+
+            if amm_amount_in.is_zero() {
+                return Ok((amount_in, None));
+            }
+
+            let amm_trade_info = Self::handle_amm_trade(
+                tx_type,
+                who.clone(),
+                market_id,
+                asset,
+                amm_amount_in,
+                BalanceOf::<T>::zero(),
+            )?;
+
+            Ok((amount_in.checked_sub_res(&amm_amount_in)?, amm_trade_info))
+        }
+
+        fn handle_amm_trade(
+            tx_type: TxType,
+            who: AccountIdOf<T>,
+            market_id: MarketIdOf<T>,
+            asset: AssetOf<T>,
+            amount_in: BalanceOf<T>,
+            min_amount_out: BalanceOf<T>,
+        ) -> Result<Option<AmmTradeOf<T>>, DispatchError> {
+            match tx_type {
+                TxType::Buy => {
+                    match T::Amm::buy(who, market_id, asset, amount_in, min_amount_out) {
+                        Ok(amm_trade) => Ok(Some(amm_trade)),
+                        Err(ApiError::SoftFailure(AmmSoftFail::Numerical)) => Ok(None),
+                        Err(ApiError::HardFailure(dispatch_error)) => Err(dispatch_error),
+                    }
+                }
+                TxType::Sell => {
+                    match T::Amm::sell(who, market_id, asset, amount_in, min_amount_out) {
+                        Ok(amm_trade) => Ok(Some(amm_trade)),
+                        Err(ApiError::SoftFailure(AmmSoftFail::Numerical)) => Ok(None),
```

Comment:
This execution path appears to be untested.

Reply:
6b3c930a47a2a8c227ccd0e4b629e0a7bc730c21